kale/embed/_mpca_torch.py

```diff
@@ -0,0 +1,230 @@
+# =============================================================================
+# Author: Shuo Zhou, szhou20@sheffield.ac.uk
+#         Haiping Lu, h.lu@sheffield.ac.uk or hplu@ieee.org
+# =============================================================================
+
+"""Multilinear Principal Component Analysis (MPCA) implementation in Pytorch
+
+Reference:
+    Haiping Lu, K.N. Plataniotis, and A.N. Venetsanopoulos, "MPCA: Multilinear
+    Principal Component Analysis of Tensor Objects", IEEE Transactions on Neural
+    Networks, Vol. 19, No. 1, Page: 18-39, January 2008. For initial Matlab
+    implementation, please go to https://uk.mathworks.com/matlabcentral/fileexchange/26168.
+"""
+# import pytorch_lightning as pl
+import warnings
+
+import tensorly as tl
+import torch
+import torch.nn as nn
+
+# from tensorly import fold, unfold
+from tensorly.tenalg import multi_mode_dot
+
+from ._mpca import _check_dim_shape
+
+tl.set_backend("pytorch")
+
+
+class MPCA(nn.Module):
+    """MPCA implementation compatible with pytorch
+
+    Args:
+        var_ratio (float, optional): Percentage of variance explained
+            (between 0 and 1). Defaults to 0.97.
+        max_iter (int, optional): Maximum number of iteration. Defaults to 1.
+
+
+    Attributes:
+        factors (list of array-like): A list of transposed projection matrices, shapes (P_1, I_1), ...,
+            (P_N, I_N), where P_1, ..., P_N are output tensor shape for each sample.
+        idx_order (tensor): The ordering index of projected (and vectorised) features in decreasing variance.
+        # mean_ (tensor): Per-feature empirical mean, estimated from the training set, shape (I_1, I_2, ..., I_N).
+        shape_in (tuple): Input tensor shapes, i.e. (I_1, I_2, ..., I_N).
+        shape_out (tuple): Output tensor shapes, i.e. (P_1, P_2, ..., P_N).
+    """
+
+    def __init__(self, var_ratio=0.97, max_iter=1):
+        super().__init__()
+
+        self.var_ratio = var_ratio
+        if max_iter > 0 and isinstance(max_iter, int):
+            self.max_iter = max_iter
+        else:
+            raise ValueError(
+                "Number of max iterations must be an integer and greater " "than 0, but given %s" % max_iter
```

Comment:
why "  " between greater and than?
To me, this message can be more compact to remain in one line of max 120. Just say *positive integer", which means an integer greater than 0.

Reply:
> 
> 
> > * I can hardly say it is ready but I do need your comments for _mpca_torch.py.
> 
> Will do for `_mpca_torch.py` but listing pros and cons in the description for each will help.
> 
> > * It is weired that I merged the upstream/master into my branch, but pre-commit on my local machine changed the 'mixed line ending' for every single file. I am confused so I did not commit the change for passing CI checks.
> 
> If _every single file_ means files you did not make changes, your branch may not be synced with the master so you should update from master because the master has changed by pre-commit earlier by me.
> 
> Overview of your branch shows 102 commits ahead of master. There is some problem with the branch, likely an update problem.

It is 102 commits ahead of my own master branch (sz144/pykale: master) and only 9 commits ahead of the upstream master branch (pykale/pykale: master). I tried to merge the latest pykale master branch into my own sz144 master, and then run pre-commit locally. It was weired that almost every single file was changed for "mixed line ending". I am not sure whether it is an issue. I asked Xianyuan but he does not have the same problem.

Reply:
Thanks for reporting this. I found it a bit annoying sometimes but just 1 or 2 files for me rather than all files. See the [documentation for mixed line ending hook](https://gdevops.gitlab.io/tuto_git/tools/pre-commit/repos_hooks/repo_pre_commit_hooks/mixed-line-ending/mixed-line-ending.html) Possible solutions
- This may be the cause but I am not sure. We can consider modifying the options to see whether this problem will be gone. In .pre-commit-config.yaml, remove "args: [--fix=lf]" under "-   id: mixed-line-ending". You may try locally. I have pushed this change to #50 so it will be in master when merged (latest by Thursday after meeting Bob).
- Remove the mixed-line-ending hook. It does not seem to be very important in modern days.
- This may be a VS code problem (Xianyuan is using PyCharm). See https://developercommunity.visualstudio.com/idea/363894/mixed-and-inconsistent-line-endings.html and https://developercommunity.visualstudio.com/idea/888411/vs2019-itself-creates-files-with-mixed-line-ending.html